### PR TITLE
Add support for additional enhanced metrics to Lambda Extension; Refactor

### PR DESCRIPTION
**Pull request by nhinsch**
### What does this PR do?

- Add support for the following enhanced metrics:
  - Timeouts
  - Out of Memory
  - Estimated Cost
- Tag enhanced metrics with the `function_arn`
- Refactor enhanced metrics-related code into its own functions/file and add unit tests

### Motivation

Bringing the Lambda Extension towards feature parity with the Datadog Forwarder.

### Describe your test plan

Unit tests are included in the PR, and I have also deployed the extension with test functions to verify the behavior.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -174,6 +174,7 @@ require (
 	gopkg.in/ini.v1 v1.55.0 // indirect
 	gopkg.in/yaml.v2 v2.2.8
 	gopkg.in/zorkian/go-datadog-api.v2 v2.29.0
+	gotest.tools v2.2.0+incompatible
```
Comment (DarcyRaynerDD) [inline go.mod:177]: Do we know why this dep was added?

File: pkg/serverless/enhanced_metrics_test.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import (
+	"testing"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/serverless/aws"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGenerateEnhancedMetricsFromFunctionLogOutOfMemory(t *testing.T) {
```
Comment (DarcyRaynerDD) [inline pkg/serverless/enhanced_metrics_test.go:12]: These unit tests are great

File: pkg/serverless/protocol.go [hunk 3]
```diff
@@ -111,59 +110,33 @@ func (l *LogsCollection) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		log.Error("Can't read log message")
 		w.WriteHeader(400)
 	} else {
+		metricsChan := l.daemon.aggregator.GetBufferedMetricsWithTsChannel()
```
Comment (DarcyRaynerDD) [inline pkg/serverless/protocol.go:113]: I like that we split all this logic out into it's own file.

**Review by DarcyRaynerDD — CHANGES_REQUESTED** [review]
Overall looks really good. Not sure about the added dep though.

File: pkg/logs/agent.go [hunk 4]
```diff
@@ -104,11 +106,12 @@ func NewServerless(sources *config.LogSources, services *service.Services, proce
...
-		destinationsCtx:  destinationsCtx,
-		pipelineProvider: pipelineProvider,
-		inputs:           inputs,
-		health:           health,
+		auditor:                   auditor,
+		destinationsCtx:           destinationsCtx,
+		pipelineProvider:          pipelineProvider,
+		inputs:                    inputs,
+		health:                    health,
+		diagnosticMessageReceiver: diagnosticMessageReceiver,
```
Comment (nhinsch) [inline pkg/logs/agent.go:114]: @remeh -- I added this because without it, the logs agent was crashing

**Review by DarcyRaynerDD — APPROVED** [review]

File: pkg/logs/agent.go (hunk 4 above)
Comment (remeh) [reply pkg/logs/agent.go:114]: Ok! That was the right thing to do, this is something brand new (21st of January) in `master` which is indeed needed!

**Review by remeh — APPROVED** [review]
LGTM! Thanks for the refactoring as well, they're improving the code. 👍

**Pull request merged**